### PR TITLE
Adds SDK support for scheduling a workflow to run after a source

### PR DESCRIPTION
**Pull request by saurav-c**
## Describe your changes and why you are making these changes
This PR adds support for scheduling a workflow to run based on a cascading update. The `publish_flow` API is changed to now allow users to set a `source_flow_id`. If this field is set then the cron schedule is ignored.

## Related issue number (if any)
ENG 2179

## Checklist before requesting a review
- [x] I have created a descriptive PR title. The PR title should complete the sentence "This PR...".
- [x] I have performed a self-review of my code.
- [ ] I have included a small demo of the changes. For the UI, this would be a screenshot or a Loom video.
- [ ] If this is a new feature, I have added unit tests and integration tests.
- [ ] I have run the integration tests locally and they are passing.
- [x] I have run the linter script locally (See `python3 scripts/run_linters.py -h` for usage).
- [ ] All features on the UI continue to work correctly.
- [ ] Added one of the following CI labels:
    - `run_integration_test`: Runs integration tests
    - `skip_integration_test`: Skips integration tests (Should be used when changes are ONLY documentation/UI)

### CODE REVIEW

**Comment by saurav-c** [comment]
@vsreekanti for API changes

File: sdk/aqueduct/client.py [hunk 1]
```diff
@@ -369,6 +371,10 @@ def publish_flow(
                 - engine: Specify where this flow should run with one of your connected integrations.
                 - k_latest_runs: Number of most-recent runs of this flow that Aqueduct should store.
                     Runs outside of this bound are deleted. Defaults to persisting all runs.
+            source_flow_id:
```
Comment (vsreekanti) [inline sdk/aqueduct/client.py:374]: Do we have to require to to be a UUID? Can we accept a name and convert?

**Review by vsreekanti — COMMENTED** [review]
Mostly looks good — just would prefer to not only accept UUIDs but also names.

File: sdk/aqueduct/client.py (hunk 1 above)
Comment (saurav-c) [reply sdk/aqueduct/client.py:374]: @vsreekanti Do you want 2 separate parameters:
- `source_flow_id`
- `source_flow_name`
And we use whichever one is set (using the UUID if both are present).
?

Comment (vsreekanti) [reply sdk/aqueduct/client.py:374]: Why don't we have the parameter be called `source_flow`:
1. We look to see if a workflow of that name exists. If it does, we use it.
2. If it doesn't, we try to convert it to a UUID. If it fails we error.
3. If it converts, we look for a workflow with that ID. If it doesn't exist we error; if it does, we use it.

Comment (saurav-c) [reply sdk/aqueduct/client.py:374]: Sounds good

Comment (kenxu95) [reply sdk/aqueduct/client.py:374]: Oh I didn't see Vikram's follow up comment lol, glad we came to the same name conclusion. Agreed it's just some conversion logic.

File: sdk/aqueduct/models/dag.py [hunk 2]
```diff
@@ -28,6 +28,7 @@ class Schedule(BaseModel):
     trigger: Optional[TriggerType] = None
     cron_schedule: str = ""
     disable_manual_trigger: bool = False
+    source_id: Optional[uuid.UUID] = None
```
Comment (kenxu95) [inline sdk/aqueduct/models/dag.py:31]: Can you add a comment here describing what this represents, and that it is only relevant for a specific trigger type?

File: sdk/aqueduct/utils/utils.py [hunk 3]
```diff
@@ -104,6 +110,16 @@ def parse_user_supplied_id(id: Union[str, uuid.UUID]) -> str:
     return id
 
 
+def parse_user_supplied_id_as_uuid(id: Union[str, uuid.UUID]) -> uuid.UUID:
```
Comment (kenxu95) [inline sdk/aqueduct/utils/utils.py:113]: nit: Can we name the other helper symmetrically with `parse_..._as_str()`  then?

**Review by kenxu95 — COMMENTED** [review]
Where are the integration tests?

File: sdk/aqueduct/client.py (hunk 1 above)
Comment (saurav-c) [reply sdk/aqueduct/client.py]: Addressed

**Review by kenxu95 — COMMENTED** [review]
Looks good just one comment. 

Just want to clarify will you add the integration tests with this PR? And will you be able to get that out before release? cc: @vsreekanti @cw75 I'd argue that we should properly add test cases before releasing features, but curious what ya'll think.

**Comment by saurav-c** [comment]
@kenxu95 I'm currently working on the integration tests. I should be able to have that complete within the next hour if that's fine for the release @vsreekanti.

**Comment by cw75** [comment]
Yeah sounds good to me

**Comment by saurav-c** [comment]
@kenxu95 @cw75 Integration test case for this is here https://github.com/aqueducthq/aqueduct/pull/863. (I need to rebase all of these PRs to pick up Kenny's refactoring changes.)

File: sdk/aqueduct/client.py [hunk 4]
```diff
@@ -416,6 +420,38 @@ def publish_flow(
                 "`artifacts` argument must either be an artifact or a list of artifacts."
             )
 
+        if source_flow and schedule != "":
+            raise InvalidUserArgumentException(
+                "Cannot create a flow with a schedule and a source flow, pick one."
```
Comment (kenxu95) [inline sdk/aqueduct/client.py:425]: super nit: "... both a schedule and.."

**Review by kenxu95 — APPROVED** [review]
cool, thanks!

**Pull request merged**